Пишем SQL-запросы

a-rotari opened the pull request:
Some increased VARCHAR size to 255 in schema.sql to accomodate longer names and descriptions, added the task-required SQL queries to queries.sql

---
:mortar_board: [Пишем SQL-запросы](https://up.htmlacademy.ru/php/12/user/1696597/tasks/10)

a-rotari committed:
Some increased VARCHAR size to 255 in schema.sql to accomodate longer names and descriptions, added the required SQL queries to queries.sql

Some increased VARCHAR size to 255 in schema.sql to accomodate longer names and descriptions, added the required SQL queries to queries.sql

grabatui commented on queries.sql:
```diff
@@ -0,0 +1,200 @@
+      bids
+    GROUP BY
+      lot_id
+  ) AS b ON lots.id = b.lot_id
```
Выглядит переусложнённо и сложночитаемо. Почему бы просто не выбрать `MAX(b.price) AS current_max_price` в общем селекте с обычным join'ом у bids'ов и `GROUP BY` по `b.lot_id`?

a-rotari replied:
Вот так?
![2022-04-13_19-54-42](https://user-images.githubusercontent.com/74206970/163231574-9bd4e8e7-b2f4-4e20-adf6-1fb026cc1dfd.png)

Но тогда, если я ничего не упускаю:

1. Придется группировать по всем остальным выводимым колонкам
2. Запрос не выдаст открытых лотов, у которых вообще нет ставок, потому что INNER JOIN не дает возможность получить в выборку NULL значение b.price. Т.е. второй сверху лот не попадет в выборку вообще
![2022-04-13_20-01-27](https://user-images.githubusercontent.com/74206970/163232558-a43a09bd-5cc3-42cc-b6b5-592fec5ec7be.png)

3. Запрос не выдаст лотов, у которых нет категории

Чтобы это прояснить нужна консультация? Я не очень понял, как ее брать.

grabatui replied:
Согласно [ТЗ](https://up.htmlacademy.ru/php/12/project/yeticave#yeti-4-5) категория является обязательным для заполнения полем. То есть мы должны верить, что категория у лота в бд всегда заполнена валидно. Предполагать что в бд кривые данные - скользкая дорожка, которая очень может усложнять код.
Но основной ответ на все 3 пункта - не обязательно доставать данные с помощью `INNER JOIN`'а. При использовании `LEFT JOIN`'а, если в поле данные попадать не будут, мы просто получим `null`. Это относится и к цене, и к названию категории.
Если использовать `LEFT JOIN`, то перечислять все доставаемые поля в `GROUP BY` в теории будет не нужно.

На всякий случай - под "обычным" join'ом я подразумевал не обязательно `inner`. Я подразумевал что не нужно `join` с селектом писать.

З.Ы. У тебя в схеме заметил `category_id INT NULL` у лота, что, как следствие первого абзаца, ошибка.

grabatui replied:
Если остались вопросы - пиши мне в любой указанный канал связи в Академии. Договоримся о консультации (или проведём письменную, если её будет достаточно).

a-rotari replied:
Переписал запрос и избавился от подзапроса. В GROUP BY придется перечислять довольно много колонок, но ведь это ничего? Я решил, что не нужно подставлять `initial_price` в `current_price` если `current_price IS NULL`, т.к. в задании ничего про это не сказано. Теперь если для открытого лота нет заявок, то в поле `current_price` выводится `NULL`
![2022-04-14_13-17-35](https://user-images.githubusercontent.com/74206970/163365287-e451013a-7a33-40d4-a948-8e948fa6486f.png)
.

grabatui replied:
> Если использовать LEFT JOIN, то перечислять все доставаемые поля в GROUP BY в теории будет не нужно.

Это не так?

a-rotari replied:
ну видимо нет :) у меня MySQL 5.7 и она ругается, если не упоминаю в GROUP BY все колонки из SELECT помимо функций агрегирования.

grabatui replied:
Поднял базу, выполнил твои схемы, твои запросы на добавление.

Работает без ошибок даже так :) Внимание на блок `group by`
![image](https://user-images.githubusercontent.com/15226721/163423298-4553d1f5-c313-46f3-a6b8-67e05bb2f563.png)

a-rotari replied:
Спасибо, буду знать!

grabatui commented on queries.sql:
```diff
@@ -0,0 +1,200 @@
+FROM
+  lots
+WHERE
+  lots.id = 5;
```
В ТЗ:
> показать лот по его ID. Получите также **название категории**, к которой принадлежит лот;

У тебя только название самого лота

a-rotari replied:
Пропустил по невнимательности. Исправил.

a-rotari committed:
Resolved the reviewer's comments

Rewrote the second SQL query getting rid of subquery, also got rid of the NULL check in that query; got rid of (if block) in the second query  no bids are present for a lot then current_price is displayed as NULL. ALso fixed the third query -- it now also displays the category of the selected lot as instructed in the task.

a-rotari committed:
Refactored a query

Refactored a query by shortening the GROUP BY clause using 'functional dependency detection' feature supported by  MySQL versions 5.7 and higher
